runtime/compiler/z/codegen/J9BCDTreeEvaluator.cpp

```diff
@@ -2069,11 +2069,46 @@ J9::Z::TreeEvaluator::BCDCHKEvaluatorImpl(TR::Node * node,
    for (uint32_t i = 0; i < numCallParam; ++i)
       callNode->setAndIncChild(i, childRootNode->getChild(i + callChildStartIndex));
 
-   // Evaluate secondChild's children, if the secondChild is an address node into a byte[]
-   if(isResultPD && secondChild->getNumChildren() == 2 && secondChild->getReferenceCount() > 1)
-      {
-      cg->evaluate(secondChild->getFirstChild());
-      cg->evaluate(secondChild->getSecondChild());
+   // Evaluate addressNode's children, if it is an address node into a byte[]
+   if(isResultPD
+         && (addressNode->getNumChildren() == 2 // case 1 and 3
+            || addressNode->isDataAddrPointer())) // case 2
+      {
+      /* Expected second child (address) node trees
+       * case 1: off-heap mode
+       *     aladd
+       *      aloadi  <contiguousArrayDataAddrFieldSymbol>
+       *        arrayObject
+       *      offset
+       *
+       * case 2: off-heap mode, accessing first element of the array
+       *     aloadi  <contiguousArrayDataAddrFieldSymbol>
+       *       arrayObject
+       *
+       * case 3: non off-heap mode
+       *     aladd
+       *      arrayObject
+       *      offset
+       *
+       * contiguousArrayDataAddrFieldSymbol isn't being commoned right now so it shouldn't
+       * have reference count > 1; arrayObject and offset nodes are the only nodes we need to
+       * evaluate early.
+       */
+      TR::Node *arrayObjectNode = addressNode->getFirstChild();
+      if (addressNode->getFirstChild()->isDataAddrPointer())
```

Comment:
this narrow checking will run into bugs i imagined (applicable to the outer checking too) ... i had similar comments for @rmnattas as well for his x86 change. for a compilation which may handle dataAddr load as internal pointer to begin with, i can envision the dataAddr load being turned into aRegLoad which doesn't have isDataAddrPointer property anymore ... you will run into bugs. currently at ilGen time, i know dataAddr load isn't treated as a full fledged internal pointer yet, but anything prevents hot/scorching compilations from commoning them up and making them into aRegStore and aRegLoad?

Reply:
>but anything prevents hot/scorching compilations from commoning them up and making them into aRegStore and aRegLoad?

dataAddr load would have to be marked as auto/param for it to be transformed to regStore/regLoad. Don't think that is the case at the moment.

Reply:
optimizer is supposed to make localization of them, as if GRA candidates were local temps before being transformed into aRegLoad/Store.  so, the question is:  is dataAddr load currently explicitly excluded from/disabled for localization?

Reply:
I can check, what would be the best place to verify this?
But if dataAddr can be transformed then don't similar checks on [p](https://github.com/eclipse-openj9/openj9/blob/22d04e7f62749bc2424159100f71665f67870e10/runtime/compiler/p/codegen/J9TreeEvaluator.cpp#L8315),[aarch64](https://github.com/eclipse-openj9/openj9/blob/22d04e7f62749bc2424159100f71665f67870e10/runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp#L5106) and [x](https://github.com/eclipse-openj9/openj9/blob/22d04e7f62749bc2424159100f71665f67870e10/runtime/compiler/x/codegen/J9TreeEvaluator.cpp#L3034-L3036) which use `->isDataAddrPointer()` also run the risk of failing?

Reply:
from the fact we knew: stackWalker needs to handle internal pointer(s) and there were a few GC bugs re displacement calculation. i am pretty sure dataAddr load or aladd indeed could be localized.